File: client/uptunnel/healthlog.py
"""Optional bounded health log for the agent side of the connection.

The console log tells you what is happening while you are watching. This tells
you what happened at 3am: connect, handshake, every keepalive round trip, and
how each session ended. When a tunnelled device goes unreachable the question is
always which end gave up first, and answering it needs both ends writing to a
file rather than to a terminal nobody was at.

Off unless UPTUNNEL_HEALTH_LOG names a path. Capped at
UPTUNNEL_HEALTH_LOG_MAX_LINES (default 1000), trimmed to the newest 80% — the
same shape as the Node agent's healthlog.ts and the server's health log, so one
description in the docs covers all three.
"""

import logging
import os
import time

DEFAULT_MAX_LINES = 1000
_KEEP_RATIO = 0.8

log = logging.getLogger("uptunnel")

_file = None
_max_lines = DEFAULT_MAX_LINES
_lines = 0
# Set after a write failure, so a bad path is reported once rather than per line.
_broken = False


def _count_lines(path):
    try:
        with open(path, "r", encoding="utf-8", errors="replace") as f:
            return sum(1 for _ in f)
    except OSError:
        return 0        # no file yet, which counts as empty


def configure_from_env():
    """Read the env and open the log. Safe to call when neither var is set."""
    global _file, _max_lines, _lines, _broken
    path = os.environ.get("UPTUNNEL_HEALTH_LOG") or None
    _broken = False
    try:
        _max_lines = int(os.environ.get("UPTUNNEL_HEALTH_LOG_MAX_LINES", DEFAULT_MAX_LINES))
    except ValueError:
        _max_lines = DEFAULT_MAX_LINES
    if _max_lines <= 10:
        _max_lines = DEFAULT_MAX_LINES
    _file = path
    if not _file:
        return
    try:
        parent = os.path.dirname(_file)
        if parent:
            os.makedirs(parent, exist_ok=True)
        _lines = _count_lines(_file)
    except OSError as exc:
        _broken = True
        log.warning("health log unavailable (%s)", exc)


def _trim():
    global _lines
    keep = int(_max_lines * _KEEP_RATIO)
    tmp = _file + ".tmp"
    try:
        with open(_file, "r", encoding="utf-8", errors="replace") as src:
            kept = src.readlines()[-keep:]
        # Write-then-rename, so an interrupted trim leaves the previous good log
        # rather than a half-written one.
        with open(tmp, "w", encoding="utf-8") as dst:
            dst.writelines(kept)
        os.replace(tmp, _file)
        _lines = len(kept)
    except OSError as exc:
        log.warning("health log trim failed (%s)", exc)
        _lines = _count_lines(_file)


def health(msg, **fields):
    """Record one health event. Never raises — diagnostics must not kill the agent."""
    global _lines, _broken
    if not _file or _broken:
        return
    tail = "".join(
        " %s=%s" % (k, v) for k, v in sorted(fields.items()) if v is not None
    )
    stamp = time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime()) + "Z"
    try:
        with open(_file, "a", encoding="utf-8") as f:
            f.write("%s %s%s\n" % (stamp, msg, tail))
    except OSError as exc:
        _broken = True
        log.warning("health log write failed, disabling (%s)", exc)
        return
    _lines += 1
    if _lines >= _max_lines:
        _trim()

File: client-node/src/healthlog.ts
/**
 * Bounded health log for the agent side of the connection.
 *
 * The console log tells you what is happening while you are watching. This
 * tells you what happened at 3am: connect, handshake, every heartbeat and its
 * round trip, and how each session ended. When a device is unreachable, the
 * question is always which end gave up first — that needs both ends writing to
 * a file, not just the server, so this is on by default.
 *
 * Written to UPTUNNEL_HEALTH_LOG, defaulting to ./health.log in the working
 * directory; set the variable to an empty string to turn it off. Capped at
 * UPTUNNEL_HEALTH_LOG_MAX_LINES (default 1000) with the newest 80% kept on
 * trim, matching the server's health log and the Pico's tunnel.log.
 */

import { appendFileSync, mkdirSync, readFileSync, renameSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { dirname, join } from "node:path";

import { log } from "./log.js";

const KEEP_RATIO = 0.8;
const DEFAULT_MAX_LINES = 1000;

/**
 * Relative to the working directory, the same shape as the server's
 * HEALTH_LOG_FILE: the log lands next to wherever the agent was started, which
 * is where anyone debugging it will look first. Set UPTUNNEL_HEALTH_LOG to an
 * absolute path for a service that runs from a directory you'd rather not
 * write to.
 */
export const DEFAULT_HEALTH_LOG = "health.log";

let file: string | null = null;
let maxLines = DEFAULT_MAX_LINES;
let lines = 0;
let broken = false;

function countLines(path: string): number {
  try {
    const body = readFileSync(path, "utf8");
    if (body.length === 0) return 0;
    return body.split("\n").length - (body.endsWith("\n") ? 1 : 0);
  } catch {
    return 0;
  }
}

/** Read the env and open the log. Safe to call when neither var is set. */
export function configureFromEnv(): void {
  // Unset means "use the default"; set-but-empty is the explicit opt out, the
  // same convention the server's HEALTH_LOG_FILE uses.
  const path = process.env.UPTUNNEL_HEALTH_LOG ?? DEFAULT_HEALTH_LOG;
  // Nothing expands a ~ out of an env file, and a path set there is exactly
  // where someone reaches for one, so expand it rather than creating a
  // directory literally named ~.
  const expanded = path.startsWith("~/") ? join(homedir(), path.slice(2)) : path;
  file = expanded.length > 0 ? expanded : null;
  broken = false;
  const raw = Number.parseInt(process.env.UPTUNNEL_HEALTH_LOG_MAX_LINES ?? "", 10);
  maxLines = Number.isFinite(raw) && raw > 10 ? raw : DEFAULT_MAX_LINES;
  if (!file) return;
  try {
    mkdirSync(dirname(file), { recursive: true });
    lines = countLines(file);
  } catch (err) {
    broken = true;
    log.warn(`health log unavailable (${(err as Error).message})`);
  }
}

function trim(): void {
  if (!file) return;
  const keep = Math.floor(maxLines * KEEP_RATIO);
  try {
    const all = readFileSync(file, "utf8").split("\n");
    if (all.length && all[all.length - 1] === "") all.pop();
    const tmp = `${file}.tmp`;
    // Write-then-rename so an interrupted trim cannot lose the whole log.
    writeFileSync(tmp, all.slice(-keep).join("\n") + "\n");
    renameSync(tmp, file);
    lines = Math.min(keep, all.length);
  } catch (err) {
    log.warn(`health log trim failed (${(err as Error).message})`);
    lines = countLines(file);
  }
}

/** Record one health event. Never throws — diagnostics must not kill the agent. */
export function health(msg: string, fields?: Record<string, unknown>): void {
  if (!file || broken) return;
  const tail = fields
    ? " " +
      Object.entries(fields)
        .filter(([, v]) => v !== undefined && v !== null)
        .map(([k, v]) => `${k}=${typeof v === "string" ? v : JSON.stringify(v)}`)
        .join(" ")
    : "";
  try {
    appendFileSync(file, `${new Date().toISOString()} ${msg}${tail}\n`);
  } catch (err) {
    broken = true;
    log.warn(`health log write failed, disabling (${(err as Error).message})`);
    return;
  }
  if (++lines >= maxLines) trim();
}
